libbeat/publisher/pipeline/consumer.go

```diff
@@ -18,206 +18,221 @@
 package pipeline
 
 import (
-	"errors"
 	"sync"
 
-	"github.com/elastic/beats/v7/libbeat/common/atomic"
 	"github.com/elastic/beats/v7/libbeat/logp"
+	"github.com/elastic/beats/v7/libbeat/publisher"
 	"github.com/elastic/beats/v7/libbeat/publisher/queue"
 )
 
 // eventConsumer collects and forwards events from the queue to the outputs work queue.
-// The eventConsumer is managed by the controller and receives additional pause signals
-// from the retryer in case of too many events failing to be send or if retryer
-// is receiving cancelled batches from outputs to be closed on output reloading.
+// It accepts retry requests from batches it vends, which will resend them
+// to the next available output.
 type eventConsumer struct {
 	logger *logp.Logger
-	ctx    *batchContext
 
-	pause atomic.Bool
-	wait  atomic.Bool
-	sig   chan consumerSignal
-	wg    sync.WaitGroup
+	// eventConsumer calls the observer methods eventsRetry and eventsDropped.
+	observer outputObserver
 
-	queue    queue.Queue
-	consumer queue.Consumer
+	// When the output changes, the new target is sent to the worker routine
+	// on this channel. Clients should call eventConsumer.setTarget().
+	targetChan chan consumerTarget
 
-	out *outputGroup
-}
+	// Failed batches are sent to this channel to retry. Clients should call
+	// eventConsumer.retry().
+	retryChan chan retryRequest
 
-type consumerSignal struct {
-	tag      consumerEventTag
-	consumer queue.Consumer
-	out      *outputGroup
-}
+	// Closing this channel signals consumer shutdown. Clients should call
+	// eventConsumer.close().
+	done chan struct{}
 
-type consumerEventTag uint8
+	// This waitgroup is released when this eventConsumer's worker
+	// goroutines return.
+	wg sync.WaitGroup
 
-const (
-	sigConsumerCheck consumerEventTag = iota
-	sigConsumerUpdateOutput
-	sigConsumerUpdateInput
-	sigStop
-)
+	// The queue the eventConsumer will retrieve batches from.
+	queue queue.Queue
+}
 
-var errStopped = errors.New("stopped")
+// consumerTarget specifies the output channel and parameters needed for
+// eventConsumer to generate a batch.
+type consumerTarget struct {
+	ch         chan publisher.Batch
+	timeToLive int
+	batchSize  int
+}
+
+// retryRequest is used by ttlBatch to add itself back to the eventConsumer
+// queue for distribution to an output.
+type retryRequest struct {
+	batch       *ttlBatch
+	decreaseTTL bool
+}
 
 func newEventConsumer(
 	log *logp.Logger,
 	queue queue.Queue,
-	ctx *batchContext,
+	observer outputObserver,
 ) *eventConsumer {
-	consumer := queue.Consumer()
 	c := &eventConsumer{
-		logger: log,
-		sig:    make(chan consumerSignal, 3),
-		out:    nil,
-
+		logger:   log,
+		observer: observer,
 		queue:    queue,
-		consumer: consumer,
-		ctx:      ctx,
-	}
 
-	c.pause.Store(true)
+		targetChan: make(chan consumerTarget),
+		retryChan:  make(chan retryRequest),
+		done:       make(chan struct{}),
+	}
 
 	c.wg.Add(1)
 	go func() {
```

Comment:
I see that you are starting multiple goroutines from the `eventConsumer`. What do you think about using `SafeWaitGroup` or `TaskGroup` from go-concert (https://pkg.go.dev/github.com/elastic/go-concert@v0.2.0/unison)?